api/service/coingecko/ticker.go

```diff
@@ -250,14 +140,73 @@ order by ps.timestamp asc
 	for i, t := range tickers {
 		baseLiquidityInUsd, err := s.liquidityInUsd(t)
 		if err != nil {
-			return nil, errors.Wrap(err, "TickerService.GetAll")
+			return nil, errors.Wrap(err, "tickerService.GetAll")
 		}
 		tickers[i].BaseLiquidityInPrice = baseLiquidityInUsd
 	}
 
 	return tickers, nil
 }
 
+func (s *tickerService) tickers(cond string, bindings ...string) ([]Ticker, error) {
+	query := `
+select distinct
+    p.asset0 base_currency,
+    p.asset1 target_currency,
+    sum(ps.volume0) over (partition by ps.pair_id) base_volume,
+    sum(ps.volume1) over (partition by ps.pair_id) target_volume,
+    t0.decimals base_decimals,
+    t1.decimals target_decimals,
+    first_value(ps.liquidity0_in_price) over (partition by ps.pair_id order by ps.timestamp desc) base_liquidity_in_price,
+    p.contract pool_id,
+    first_value(ps.timestamp) over (partition by ps.pair_id order by ps.timestamp desc) as timestamp
+from pair_stats_recent ps
+    join pair p on ps.pair_id = p.id
+    join tokens t0 on p.chain_id = t0.chain_id and p.asset0 = t0.address
+    join tokens t1 on p.chain_id = t1.chain_id and p.asset1 = t1.address
+where ps.chain_id = ?
+  and ps.timestamp >= extract(epoch from now()-interval'24h')
+`
+	var tickers []Ticker
+	if cond != "" {
+		b := make([]interface{}, len(bindings)+1)
+		b[0] = s.chainId
+		for i, v := range bindings {
+			b[i+1] = v
+		}
+		if tx := s.Raw(query+" and "+cond, b...).Find(&tickers); tx.Error != nil {
+			return nil, errors.Wrap(tx.Error, "tickerService.tickers")
+		}
+	} else {
+		if tx := s.Raw(query, s.chainId).Find(&tickers); tx.Error != nil {
+			return nil, errors.Wrap(tx.Error, "tickerService.tickers")
+		}
+	}
+
+	for i, t := range tickers {
+		baseDecimal := types.NewDec(10).Power(uint64(t.BaseDecimals))
+		targetDecimal := types.NewDec(10).Power(uint64(t.TargetDecimals))
```

Comment:
If you used `sdk.NewDecWithPrec` or `sdk.NewDecFromBigIntWithPrec`, it can avoid pow.

Reply:
thx 👍  1c28c5f